libcudacxx/docs/extended_api/synchronization_primitives/barrier.md

```diff
@@ -60,6 +61,43 @@ Programs shall ensure that this transformation would not introduce errors, for
 Under CUDA Compute Capability 6 (Pascal) or prior, an object of type
   `cuda::barrier` or `cuda::std::barrier` may not be used.
 
+## Shared memory barriers with transaction count
+
+Starting with the Hopper architecture (CUDA Compute Capability 9), a
+`cuda::barrier` object located in shared memory supports a new count, called
+[tx-count](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#tracking-asynchronous-operations-by-the-mbarrier-object),
+which is used for tracking the completion of asynchronous memory operations or
+transactions. The tx-count tracks the number of asynchronous transactions, in
+units specified by the asynchronous memory operation (typically bytes), that are
+outstanding and yet to be complete.
+
+The tx-count of `cuda::barrier` must be set to the total amount of asynchronous
+memory operations, in units as specified by the asynchronous operations, to be
+tracked by the current phase. This can be achieved with the
+`cuda::device::barrier_arrive_tx` function call. Upon completion of each of the
+asynchronous operations, the tx-count of the `cuda::barrier` will be updated and
+thus progress the `cuda::barrier` towards the completion of the current phase.
+This may complete the current phase.
+
+### Phase Completion of a `cuda::barrier` with tx-count support
+
+The requirements for completion of the current phase are described below. Upon
```

Comment:
I think this section needs to be worded as a diff to the phase completion step of the barrier.

Reply:
Do you mean as a diff to this section?

> A barrier phase consists of the following steps:

>    The expected count is decremented by each call to [arrive](https://en.cppreference.com/w/cpp/thread/barrier/arrive) or [arrive_and_drop](https://en.cppreference.com/w/cpp/thread/barrier/arrive_and_drop).
    When the expected count reaches zero, the phase completion step is run, meaning that the CompletionFunction is invoked, and all threads blocked on the phase synchronization point are unblocked. The end of the completion step [strongly happens-before](https://en.cppreference.com/w/cpp/atomic/memory_order#Strongly_happens-before) all calls that were unblocked by the completion step return.
    Exactly once after the expected count reaches zero, a thread executes the completion step during its call to [arrive](https://en.cppreference.com/w/cpp/thread/barrier/arrive), [arrive_and_drop](https://en.cppreference.com/w/cpp/thread/barrier/arrive_and_drop), or [wait](https://en.cppreference.com/w/cpp/thread/barrier/wait), except that it is implementation-defined whether the step executes if no thread calls [wait](https://en.cppreference.com/w/cpp/thread/barrier/wait).
    When the completion step finishes, the expected count is reset to the value specified at construction less the number of calls to [arrive_and_drop](https://en.cppreference.com/w/cpp/thread/barrier/arrive_and_drop) since, and the next barrier phase begins.

from https://en.cppreference.com/w/cpp/thread/barrier

Reply:
Right. Something along the line of what we do in the memory model section:

Modify [[thread.barrier.class]](http://eel.is/c++draft/thread.barrier.class) as follows: 

> A barrier is a thread coordination mechanism whose lifetime consists of a sequence of barrier phases, where each phase allows at most an expected number of threads to block until the expected number of threads **and the expected number of transaction-based asynchronous operations** arrive at the barrier.
>
> Each _barrier phase_ consists of the following steps:
> 1. The _expected count_ is decremented by each call to `arrive`,`arrive_and_drop`**, or `cuda::device::barrier_arrive_tx`**. 
> 2. **The _transaction count_ is incremented by each call to `cuda::device::barrier_arrive_tx` and decremented by the completion of transaction-based asynchronous operations such as `cuda::memcpy_async_tx`.**
> 3. Exactly once after **both** the _expected count_ **and the _transaction count_** reach zero, a thread executes the _completion step_ during its call to `arrive`, `arrive_and_drop`, or `wait`, except that it is implementation-defined whether the step executes if no thread calls `wait`.
> 4. When the completion step finishes, the _expected count_  is reset to what was specified by the `expected` argument to the constructor, possibly adjusted by calls to `arrive_and_drop`, **the _transaction count_ is reset to zero,** and the next phase starts.
>
> Concurrent invocations of the member functions of barrier **and the non-member barrier APIs in `cuda::device`**, other than its destructor, do not introduce data races. The member functions `arrive` and `arrive_and_drop`, **and the non-member function `cuda::device::barrier_arrive_tx`**, execute atomically.

Reply:
Thanks. I have replaced the current text with your proposal.